internal/transport/grpchttp2/http2bridge.go

```diff
@@ -0,0 +1,219 @@
+// alongside the maxHeaderListSize for the maximum size of the headers table.
+// The underlying framer uses the SetReuseFrames feature to avoid extra
+// allocations.
+func NewFramerBridge(w io.Writer, r io.Reader, maxHeaderListSize uint32) *FramerBridge {
+	fr := &FramerBridge{
+		framer: http2.NewFramer(w, r),
+		pool:   mem.DefaultBufferPool(),
+	}
+
+	fr.framer.SetReuseFrames()
+	fr.framer.MaxHeaderListSize = maxHeaderListSize
+	fr.framer.ReadMetaHeaders = hpack.NewDecoder(initHeaderTableSize, nil)
+
+	return fr
+}
+
+// ReadFrame reads a frame from the underlying http2.Framer and returns a
+// Frame defined in the grpchttp2 package.
+func (fr *FramerBridge) ReadFrame() (Frame, error) {
+	f, err := fr.framer.ReadFrame()
+	if err != nil {
+		return nil, err
+	}
+
+	h := f.Header()
+	hdr := &FrameHeader{
+		Size:     h.Length,
+		Type:     FrameType(h.Type),
+		Flags:    Flag(h.Flags),
+		StreamID: h.StreamID,
+	}
+
+	switch f := f.(type) {
+	case *http2.DataFrame:
+		buf := fr.pool.Get(int(hdr.Size))
+		copy(buf, f.Data())
+		df := &DataFrame{
+			hdr:  hdr,
+			Data: buf,
+		}
+		df.free = func() { fr.pool.Put(buf) }
+		return df, nil
+	case *http2.HeadersFrame:
+		buf := fr.pool.Get(int(hdr.Size))
+		copy(buf, f.HeaderBlockFragment())
+		hf := &HeadersFrame{
+			hdr:      hdr,
+			HdrBlock: buf,
+		}
+		hf.free = func() { fr.pool.Put(buf) }
+		return hf, nil
+	case *http2.RSTStreamFrame:
+		return &RSTStreamFrame{
+			hdr:  hdr,
+			Code: ErrCode(f.ErrCode),
+		}, nil
+	case *http2.SettingsFrame:
+		buf := make([]Setting, 0, f.NumSettings())
+		f.ForeachSetting(func(s http2.Setting) error {
+			buf = append(buf, Setting{
+				ID:    SettingID(s.ID),
+				Value: s.Val,
+			})
+			return nil
+		})
+		return &SettingsFrame{
+			hdr:      hdr,
+			Settings: buf,
+		}, nil
+	case *http2.PingFrame:
+		buf := fr.pool.Get(int(hdr.Size))
+		copy(buf, f.Data[:])
+		pf := &PingFrame{
+			hdr:  hdr,
+			Data: buf,
+		}
+		pf.free = func() { fr.pool.Put(buf) }
+		return pf, nil
+	case *http2.GoAwayFrame:
+		// Size of the frame minus the code and lastStreamID
+		buf := fr.pool.Get(int(hdr.Size) - 8)
+		copy(buf, f.DebugData())
+		gf := &GoAwayFrame{
+			hdr:          hdr,
+			LastStreamID: f.LastStreamID,
+			Code:         ErrCode(f.ErrCode),
+			DebugData:    buf,
+		}
+		gf.free = func() { fr.pool.Put(buf) }
+		return gf, nil
+	case *http2.WindowUpdateFrame:
+		return &WindowUpdateFrame{
+			hdr: hdr,
+			Inc: f.Increment,
+		}, nil
+	case *http2.ContinuationFrame:
+		buf := fr.pool.Get(int(hdr.Size))
+		copy(buf, f.HeaderBlockFragment())
+		cf := &ContinuationFrame{
+			hdr:      hdr,
+			HdrBlock: buf,
+		}
+		cf.free = func() { fr.pool.Put(buf) }
+		return cf, nil
+	case *http2.MetaHeadersFrame:
+		return &MetaHeadersFrame{
+			hdr:    hdr,
+			Fields: f.Fields,
+		}, nil
+	}
+
+	return nil, connError(ErrCodeProtocol)
+}
+
+func (fr *FramerBridge) WriteData(streamID uint32, endStream bool, data ...[]byte) error {
+	var buf []byte
```

Comment:
Optionally, if we specify that `WriteData` must not be called with `0` byte slices, we could re-organize this as follows:
``` golang
buf := data[0]
if len(data) != 1 {
  // The whole logic you currently have of computing the total length and copying over the data
}
return fr.framer.WriteData(streamID, endStream, buf)
```

Reply:
I forgot to reply to this conversation. I think that could work, but I remember `loopyWriter` [does send DataFrames with `nil` slices ](https://github.com/printchard/grpc-go/blob/21e2fa479f31c4a4b6a57aa30f7b14118cec6f5a/internal/transport/controlbuf.go#L926)(esentially empty DataFrames). So I think it might be worth to leave the behaviour as is. What do you think?

Reply:
In that case, how about?

```go
	if len(data) == 1 {
		return fr.framer.WriteData(streamID, endStream, data[0])
	}

	var buf []byte	
	tl := 0
	for _, s := range data {
		tl += len(s)
	}

	buf = fr.pool.Get(tl)[:0]
	defer fr.pool.Put(buf)
	for _, s := range data {
		buf = append(buf, s...)
	}

	return fr.framer.WriteData(streamID, endStream, buf)
```

Reply:
Done.